### PR TITLE
ApiManagement-DeveloperServicesAPI 2022-04-01-preview

### LINKED ISSUE #22487: [API Management] API Review
New API Review meeting has been requested.

**Service Name**: API Management
**Review Created By**: Harun Heybet
**Review Date**: 3/7/2023 1:00 PM PT
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/22470
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: [here](https://microsofteur-my.sharepoint.com/:p:/g/personal/harunheybet_microsoft_com/Ebajf5hArddMrxvZv9FUxJcB1Q54ak7pJ25lzMplJ6WcMQ?e=Pl6Kfl)

**Description**: Developer Services API Preview Release.
https://microsofteur-my.sharepoint.com/:p:/g/personal/harunheybet_microsoft_com/Ebajf5hArddMrxvZv9FUxJcB1Q54ak7pJ25lzMplJ6WcMQ?e=Pl6Kfl

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=f801baae-2aa6-ed11-aad1-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by harunrst**
# Data Plane API - Pull Request - Developer Services API
Introducing a new data plane API for [API Management Developer Portal](https://github.com/Azure/api-management-developer-portal) customers.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/22487

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document: Internal Document.
* Previous Open API Doc: None.
* Updated paths: None.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/22487
fix https://github.com/Azure/azure-rest-api-specs/issues/22487

### CODE REVIEW

File: specification/apimanagement/developer-services-api/authorized/resources/users.tsp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    @doc("User identifier.")
+    @path
+    userId: string
+  ): NoContentResponse | ErrorResponse;
+
+  @route("/users/{userId}/token/revoke")
+  @doc("Revoke the token of a user by its identifier.")
+  @post
+  RevokeToken(
```
Comment (mikekistler) [inline specification/apimanagement/developer-services-api/authorized/resources/users.tsp:54]: This should be implemented using an operation template from Azure Core, probably `ResourceAction`.

File: specification/apimanagement/developer-services-api/authorized/resources/confirmations.tsp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using Azure.Core.Foundations;
+using DeveloperServicesApi.Models;
+
+namespace DeveloperServicesAuthorizedApi.Resources.Confirmations;
+
+interface Confirmations {
+  @route("/confirmations/users/{userId}/identities/basic")
+  @doc("Basic identity (Email) confirmation for a User.")
+  @post
+  CreateUserBasicIdentityConfirmation(
```
Comment (mikekistler) [inline specification/apimanagement/developer-services-api/authorized/resources/confirmations.tsp:17]: This should be implemented with an operation template from Azure Core, either `ResourceAction` or `RpcOperation`.

File: specification/apimanagement/developer-services-api/authorized/resources/identities.tsp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  @doc("Identity of current user.")
+  getIdentity is ResourceRead<IdentityResource>;
+
+  @doc("List a collection of user identities.")
+  listIdentities is ResourceList<UserIdentityResource>;
+
+  @route("/users/{userId}/identities/{provider}/{identity}")
+  @doc("Creates or updates the given user identity.")
+  @put
+  CreateOrUpdateUserIdentityProvider(
```
Comment (mikekistler) [inline specification/apimanagement/developer-services-api/authorized/resources/identities.tsp:24]: This should be implemented with an Azure Core operation template, perhaps `ResourceCreateOrReplace`.

File: specification/apimanagement/developer-services-api/models/example.tsp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@doc("Payload or parameter example.")
+model Example {
+  @doc("Short description for the example.")
+  summary: string;
+
+  @doc("Long description for the example.")
+  description: string;
+
+  @doc("Embedded literal example. The `value` field and `externalValue` field are mutually exclusive. To represent examples of media types that cannot naturally represented in JSON or YAML, use a string value to contain the example, escaping where necessary.")
+  value: object;
```
Comment (mikekistler) [inline specification/apimanagement/developer-services-api/models/example.tsp:18]: The `object` type is being deprecated. Should this be `unknown` (meaning "any")?

**Review by mikekistler — COMMENTED** [review]
I built using the latest TypeSpec compiler / Azure Core and it flagged a number of issues.  I've commented on some of these to give a flavor and you can find the rest by updating your TypeSpec packages.

File: specification/apimanagement/developer-services-api/models/common.tsp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+alias CreatedApiResponse<T> = Created<T> | ErrorResponse;
+
+@doc("Collection object for listing operations.")
+model Collection<T> {
+  @doc("Relative URL to next page of collection.")
+  nextLink: string;
+
+  @doc("Collection itself.")
+  value: T[];
+}
```
Comment (mikekistler) [inline specification/apimanagement/developer-services-api/models/common.tsp:26]: You should use the models from Azure Core. Reimplementing these common types only creates opportunities for inadvertent errors, as is the case here -- `nextLink` should be optional.

File: specification/apimanagement/developer-services-api/authorized/resources/users.tsp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+using DeveloperServicesApi.Models;
+using DeveloperServicesAuthorizedApi.Common;
+
+namespace DeveloperServicesAuthorizedApi.Resources.Users;
+
+interface Users {
+  @route("/users/{userId}")
+  @doc("Get a user profile by its identifier.")
+  @get
+  GetUser(
```
[the remark this answers is not in the document]
Comment (tjprescott) [reply specification/apimanagement/developer-services-api/authorized/resources/users.tsp]: @harunst this was deemed "by design" (https://github.com/Azure/typespec-azure/issues/2831) so Mike is right--this should be a ResourceRead. The fact that the password appears in the generated autorest-flavored Swagger is a scenario that autorest code generators have been designed to handle.

**Review by mikekistler — APPROVED** [review]
I still have concerns about the duplication of paths and aliasing of resources, but I'm willing to let this go forward in preview so the team can get customer feedback on this design.

File: eng/pipelines/typespec-ci.yml [hunk 7]
```diff
@@ -43,6 +44,11 @@ jobs:
       Folder: specification/contosowidgetmanager/Contoso.WidgetManager
       DisplayName: Contoso.WidgetManager
 
+  - template: templates/steps/typespec-ci.yml
```
Comment (mikeharder) [inline eng/pipelines/typespec-ci.yml:47]: @harunrst: This change adds your new spec to our CI validation pipeline.  If you rename or move the folder please update it here.

File: specification/apimanagement/developer-services-api/tspconfig.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+parameters:
```
Comment (raych1) [inline specification/apimanagement/developer-services-api/tspconfig.yaml:1]: @harunrst , please refer to [the latest sample project](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml) to correct the schema in order to generate the sdk package successfully.

**Pull request closed without merging**